crates/cargo-contract/src/cmd/extrinsics/balance.rs

```diff
@@ -0,0 +1,289 @@
+
+        if let Some(token_metadata) = token_metadata {
+            if n == 0 {
+                return BalanceVariant::Denominated(format!("0{}", token_metadata.symbol))
+            }
+
+            let units_result = n / token_metadata.denomination;
+            let mut symbol = "";
+            let remainder: u128;
+            let units: u128;
+            if (1..1_000).contains(&units_result) {
+                remainder = n % token_metadata.denomination;
+                units = units_result;
+            } else if (1_000..1_000_000).contains(&units_result) {
+                remainder = units_result % 1_000;
+                units = units_result / 1_000;
+                symbol = "k";
+            } else if (1_000_000..1_000_000_000).contains(&units_result) {
+                remainder = units_result % 1_000_000;
+                units = units_result / 1_000_000;
+                symbol = "M";
+            } else if n / 1_000_000_000 > 0 {
+                remainder = n % 1_000_000_000;
+                units = n / 1_000_000_000;
+                symbol = "n";
+            } else if n / 1_000_000 > 0 {
+                remainder = n % 1_000_000;
+                units = n / 1_000_000;
+                symbol = "μ";
+            } else {
+                remainder = n % 1_000;
+                units = n / 1_000;
+                symbol = "m";
+            }
+            if remainder > 0 {
+                let remainder = remainder.to_string().trim_end_matches('0').to_owned();
+                BalanceVariant::Denominated(format!(
+                    "{}.{}{}{}",
+                    units, remainder, symbol, token_metadata.symbol
+                ))
+            } else {
+                BalanceVariant::Denominated(format!(
+                    "{}{}{}",
+                    units, symbol, token_metadata.symbol
+                ))
+            }
+        } else {
+            BalanceVariant::Default(n)
+        }
+    }
+}
+
+impl Display for BalanceVariant {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            BalanceVariant::Default(balance) => f.write_str(&balance.to_string()),
+            BalanceVariant::Denominated(input) => f.write_str(input),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::cmd::extrinsics::parse_balance;
+
+    use super::*;
+
+    #[test]
+    fn correct_balances_parses_success() {
+        assert!(
+            parse_balance("500DOT").is_ok(),
+            "<500DOT> was not parsed correctly"
+        );
+        assert!(
+            parse_balance("500").is_ok(),
+            "<500> was not parsed correctly"
+        );
+    }
+
+    #[test]
+    fn incorrect_balances() {
+        assert!(
+            parse_balance("500%").is_err(),
+            "expected to fail parsing incorrect balance"
+        );
+    }
+
+    #[test]
+    fn balance_variant_denominated_success() {
+        let tm = TokenMetadata {
+            denomination: 12,
```

Comment:
I know `denomination` is not used in the parsing but this is an invalid value for this, it seems that in other cases it is `1_000_000_000_000`